### PR TITLE
Bitcrazed/enable windows vt support - fixes #3008

### LINKED ISSUE #3008: Enable curl to work as expected on Windows
@bagder and @jay & the curl team:

Hi - I am the PM for Windows Console. Since adding [curl to Windows 10 in the Windows 10 Sprint 2018 Update](https://blogs.msdn.microsoft.com/commandline/2018/01/18/tar-and-curl-come-to-windows/), we've seen and received reports of issue with Curl not behaving as expected on Windows, e.g. curl issues #731 and #345, and [known-issue 5.5](https://github.com/curl/curl/blob/master/docs/KNOWN_BUGS#L353)

I and the Windows Console Team, are in the process of overhauling Windows Console.

For example, we recently [added a Pseudo Console (ConPTY) API to Windows for the first time](https://blogs.msdn.microsoft.com/commandline/2018/08/02/windows-command-line-introducing-the-windows-pseudo-console-conpty/). This API was inspired by Linux' `openpty(3)` APIs and enables terminal/server applications to create and supply (via a call to [`CreatePseudoConsole()`](https://docs.microsoft.com/en-us/windows/console/closepseudoconsole)) the pipes via which they communicate directly with command-line apps (e.g. Cmd, PowerShell, curl.exe, etc.), using UTF-8 encoded text/VT.

We're also in the process of overhauling how the Console's internal buffers store and handle Unicode, and UTF-8. In the up-coming Windows 10 Fall 2018 release, Console will be able to store and retrieve Unicode text in codepage 65001 (though it'll still struggle to display emoji, chars outside the currently selected font, etc. - that's a whole other set of issues 😜) :

image

To accomplish the above, I [hacked a couple of quick changes](https://github.com/bitcrazed/curl/blob/bitcrazed/enable-windows-vt-support/src/tool_main.c#L254-L305) into my own fork of curl to enable VT-mode in the Console. when curl starts-up.

However, these changes are far from complete:

1. I am sure you'd want such changes implemented in a different way?!
1. I haven't yet figured out is where to add code that changes the codepage to 65001 at startup on Windows 10? Alas, without switching to codepage 65001 for UTF-8 HTTP responses, Console garbles its output:

image

Could you steer me towards the changes you'd recommend to enable curl to light-up on Windows?

### PR DESCRIPTION
**Pull request by bitcrazed**
As per #3008, this PR implements the changes necessary to make curl work as expected on Windows 10 that now have enough ANSI/VT & UTF-8 support to handle many more related scenarios.

First-time curl submitter, but don't be gentle - would appreciate any and all constructive criticism, change requests, and suggestions 😀

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -56,3 +56,5 @@ scripts/_curl
 curl_fuzzer
 curl_fuzzer_seed_corpus.zip
 libstandaloneengine.a
+.vscode/
+*.debugtarget
```
Comment (danielgustafsson) [inline .gitignore]: Being development environment specific, these two are probably better handled locally in `$GIT_DIR/info/exclude`

**Comment by jay** [comment]
The winbuild changes and printfs need to go. MSVTMODE macro seems unnecessary. In libcurl we use define WIN32 macro on Windows so I think `#ifdef WIN32` should be sufficient. What is the use case for changing the input?

**Comment by gvanem** [comment]
Doesn't `GetConsoleMode()` tell whether `ENABLE_VIRTUAL_TERMINAL_PROCESSING` is really sticky? 
Your change seems to split out ANSI-sequences only (and always) when `MSVTMODE` is defined. Not nice if VT-mode is not in effect. Does Win-10 support the *Device status report*?

*  `printf("\x1b[6n\r      \r")` and verify the cursor-position with `scanf("\e[%d;%dR",&y,&x)==2)`.

File: src/tool_main.c [hunk 2]
```diff
@@ -245,6 +313,11 @@ int main(int argc, char *argv[])
   struct GlobalConfig global;
   memset(&global, 0, sizeof(global));
 
+  TerminalSettings ts;
```
Comment (gvanem) [inline src/tool_main.c]: Data here is legal only in C99 (as shown in AppVeyor).

**Comment by bitcrazed** [comment]
Thanks for your feedback all.

I've made several changes to address your points and checked that curl builds and runs correctly on RS5 builds of Windows 10.

Let me know if you need anything else changed.

**Comment by bagder** [comment]
As seen on travis:
~~~
tool_main.c:240:1: error: C++ style comments are not allowed in ISO C90 [-Werror]
 // TerminalSettings for Windows
 ^
tool_main.c:240:1: error: (this will be reported only once per input file) [-Werror]
tool_main.c:248:6: error: function declaration isn’t a prototype [-Werror=strict-prototypes]
 void configure_terminal()
      ^
tool_main.c: In function ‘configure_terminal’:
tool_main.c:248:6: error: old-style function definition [-Werror=old-style-definition]
tool_main.c: At top level:
tool_main.c:277:6: error: function declaration isn’t a prototype [-Werror=strict-prototypes]
 void restore_terminal()
      ^
tool_main.c: In function ‘restore_terminal’:
tool_main.c:277:6: error: old-style function definition [-Werror=old-style-definition]
cc1: all warnings being treated as errors
~~~
Private functions should be `static` and no argument should be specified as `(void)`...

**Comment by bitcrazed** [comment]
Thanks @bagder - just saw that and have fixed those issues.

**Comment by danielgustafsson** [comment]
checksrc found a few trivial nits in the last version:
```
/usr/bin/perl ../lib/checksrc.pl -D.        \
	-W./tool_hugehelp.c ./*.[ch]
./tool_main.c:262:78: warning: Trailing whitespace (TRAILINGSPACE)
   /* Cache current codepage (will restore on exit) & set codepage to UTF-8 */ 
                                                                              ^
./tool_main.c:272:3: warning: badly placed open brace (BRACEPOS)
   {
   ^
./tool_main.c:283:68: warning: Trailing whitespace (TRAILINGSPACE)
   /* Restore Console output mode and codepage to whatever they were 
                                                                    ^
```

**Comment by bitcrazed** [comment]
Man alive! Wish these rules were turned-on by default for local builds too 😖
Of course, it doesn't help that I also accidentally deleted a character from the front of a function name ;)

Should be fixed now. Hopefully!

**Comment by bitcrazed** [comment]
WOOT! Green at last!! 😛

**Comment by bagder** [comment]
> Wish these rules were turned-on by default for local builds too

Indeed. They can be switched on for some builds (`configure --enable-debug`) but we could certainly improve by doing it for more. `make checksrc` is sometimes a quick work-around.

(there are a little too many different ways to build curl... :grimacing:)

**Comment by bagder** [comment]
Thanks a lot! You might note that I squashed the commits and edited some minor details before I pushed.

**Pull request closed without merging**